platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc

```diff
@@ -102,6 +103,33 @@ We don't expect too many occurrences given the similar nullability semantics.
 Note that plugins using JSR-305 annotations on their types will continue to work as they did with previous Gradle versions.
 In other words, it is still supported to use `javax.annotations` for nullability on your plugin types such as tasks or extensions.
 
+==== Upgrade to Kotlin 2.1.20
+
+The embedded Kotlin has been updated from 2.0.21 to link:https://github.com/JetBrains/kotlin/releases/tag/v2.1.20[Kotlin 2.1.20].
+
+Kotlin 2.1 removes support for targeting Kotlin language 1.4 and 1.5, it also deprecates using Kotlin language 1.7.
+See the Kotlin release notes for other potential breaking changes.
+
+==== The Kotlin DSL now uses the Kotlin language version 2.1
+
+Gradle 8.x used the Kotlin language version 1.8.
+Gradle 9.0 uses the Kotlin language version 2.1.
+See the Kotlin release notes for a description of the language changes and available new language features.
+
+The Kotlin 2 compiler makes change in the way scripts are compiled.
+It is not possible to refer to the script object using `this@Build_gradle`, `this@Settings_gradle` or `this@Init_gradle` anymore.
+You must prefer using the target object instead, respectively `project`, `settings` and `gradle`.
+
+Kotlin 2.1 combined with link:#jspecify[JSpecify] brings some new strictness in how nullability is handled when consuming the Gradle API from the Kotlin DSL.
+Here are some potential breaking changes.
+
+Unbounded generics for types that have generic bounds will now fail to compile.
+For example if you have a Kotlin extension function on `Provider<T>` whose signature is `fun <T> Provider<T>.some()` you must qualify `<T>` as either  `<T : Any>` or `<T : Any?>` depending on the nullability you need.
```

Comment:
Sorry for chiming in late here

❓ `T` isn't nullable for the `Provider<T>`, is `<T : Any?> Provider<T>.some()` even a thing?

Reply:
Good catch, thanks! I reworded.